---
title: "Adding Signatures to the TBSignatureProfiler"
author: 
- name: Aubrey Odom
  affiliation: 
  - Program in Bioinformatics, Boston University, Boston, MA
  email: aodom@bu.edu

package: TBSignatureProfiler
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{"Introduction to the TBSignatureProfiler"}
  %\VignetteEncoding{UTF-8}  
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

# Introduction

Thank you for helping to contribute to the TBSignatureProfiler's longitevity by adding signatures to our package. This tutorial serves as a step-by-step walkthrough to add signatures by performing the necessary code changes yourself and then pulling in the updated package into the official version in the [compbiomed repository](www.github.com/compbiomed/tbsignatureprofiler).

# Setup

Please be sure to download the latest version of the TBSignatureProfiler from [github](www.github.com/compbiomed/tbsignatureprofiler) onto your local machine. Navigate to that folder and set it as your working directory. We will use devtools to load the package.
```{r setup}
# plyr will be used later
library(plyr)

if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")

devtools::load_all()
```

# Adding Signatures


  To illustrate the process of adding signatures, we will use the simple 3-transcript signature of daCosta_OD_3 as an example. 

  ## Gathering information
  
  The first step in this process is gathering the necessary information about the signature(s) that you would like to add. For each signature, you will need to access the signature's associated publication, and make a note (a table will be helpful) of these crucial pieces of information:
  
  * *The gene transcripts composing the signature:* Currently, all signatures in the package are stored using gene symbols (not Entrez or Ensembl IDs, although these may be implemented in the future). The gene symbols composing a signature are usually either listed in the body of the article, printed in a figure, or available as a table in the supplementary materials. Sometimes the symbols aren't readily available - in which case, you can either try to contact the author or just move on from that signature.
  * *The length of the signature:* Please count the number of gene symbols in the signature - this will be used in naming the signature according to package naming standards.
  * *The name of the first author listed:* Take note of the last name of the first author listed - this will be used in naming the signature according to package naming standards.
  * *Signature type:* We need to know what the signature is intended to discriminate against. The signature type indicates whether the signature was developed to distinguish TB from LTBI ("Disease"), TB from some combination of other diseases and possibly LTBI ("Disease/Other Diseases"), TB from Human Immunodeficiency Virus ("Disease/HIV"), TB from pneumonia ("Disease/Pneumonia"), or identify risk of progression to TB ("risk"), risk of TB treatment failure ("failure"), or classify treatment responses (i.e., failures from cures, "response"). Whatever the designation, please record it as listed in parentheses above (i.e., "risk", "Disease/Other Diseases"). Note that these designations are case sensitive when we store them in the signature annotation object in the package. More details on this will be explained later in the tutorial.
  * *Tissue type:* The tissue type variable denotes whether the signature was developed using samples of either whole blood/paxgene or peripheral blood mononuclear cells (PBMCs). Due to the manipulation of cells inherently required to obtain PBMCs, many scientists prefer to use only whole blood samples for analysis. Also, note that "peripheral blood" without the peripheral blood mononuclear cell designation usually refers to whole blood, so be sure to record it correctly. If you are unsure what the tissue type is, get a second opinion to ensure correctness. 
  * *Assigned publication signature name:* Sometimes authors name their signatures in originating publication (or that of a peer). If this is the case, take note of this common name. We will use this name in the package alongside a name that we develop according to our own nomenclature system for convenience. Examples of alternative names include (Risk6)[https://pubmed.ncbi.nlm.nih.gov/32451443/], (PREDICT29)[https://www.sciencedirect.com/science/article/pii/S1472979219303907?via%3Dihub], and (ACS_COR)[https://www.sciencedirect.com/science/article/pii/S0140673615013161?via%3Dihub]

At this point, you should have recorded all of the neccessary information. If you are missing any of the pieces of information listed above, then you will not be able to add the signature at this time. If you have any concerns about recording the correct information, you can reach out to the package maintainer, Aubrey Odom, at <aodom@bu.edu>.

The final step before introducing the signature into the package is to give it a proper name, according to our package naming conventions. This is separate than the common name that may have already been given to a signature in a publication. As an example, we will use 
  
  * Should be shortened if possible - how many characters is longest?
  * Details on naming criteria...


* Table of possible things to collect info on

  ## TBsignatures list objects
    
  The first objects that need to be updated are the lists of signatures themselves. 
  
  ### Regular object
  
This is the object that most people will be using.
  
```{r list objs}

# Load current sigs -----------------------------------------------------------
data("TBsignatures")
data("TBcommon")

# Create new sig --------------------------------------------------------------

# Dawany
hiv <- read.csv("C:/Users/aubre/Documents/BU Research/Tuberculosis/TBSP Development/Dawany_HIV_251.csv")[, 2]
ind <- which(hiv == "NaN")
TBsignatures$`Dawany_HIV_251` <- hiv[-ind]

# Darboe
TBsignatures$`Darboe_RISK_11` <-  c("BATF2", "ETV7", "FCGR1B", "GBP1",  "GBP2", "GBP5", "SCARF1", "SERPING1", "STAT1", "TAP1", "TRAFD1" )

# Penn-Nicholson RISK6
TBsignatures$`PennNich_RISK_6` <- c("SERPING1", "TRMT2A", "GBP2", "SDR39U1", "FCGR1B", "TUBGCP6")

# Suliman Risk 2
TBsignatures$`Suliman_RISK_2` <- c("ANKRD22", "OSBPL10") 

# Create index of names in proper order----------------------------------------
new_names <- sort(names(TBsignatures))

# Order by sorted names
TBsignatures <- TBsignatures[new_names]

# Save new object
save(TBsignatures, file = "data/TBsignatures.rda")
```
  * Note that we will add documentation later

  
  ### Common name object
  
  Need to use the alternative name if present or the actual name otherwise.
  
```{r common name list}
  TBcommon <- TBsignatures

index <- which(names(TBcommon) %in% c("Leong_RISK_29", "Maertzdorf_4", "Suliman_RISK_4", "Sweeney_OD_3", "Thompson_9",
                             "Thompson_FAIL_13", "Thompson_RES_5", "Zak_RISK_16", "PennNich_RISK_6"))
names(TBcommon)[index] <- c("PREDICT29", "DIAG4", "RISK4", "DIAG3", "DISEASE",
                            "FAILURE", "RESPONSE5", "ACS_COR", "RISK6")
names(TBcommon)
save(TBcommon, file = "data/TBcommon.rda")
  
```
  
  ### Documentation (Do not forget this step!)
  
  The documentation is the only place where the source of the signature will be mentioned in the package; therefore, it is crucial that the documentation be updated accordingly.
  * Where to access the documentation
  * Specific format?

  ## TBsignatures objects
  
  We also need to fix up the descriptive objects... what are they used for...
  
  The disease variable indicates whether the signature was developed to distinguish TB from LTBI ("Disease"), TB from some combination of other diseases and possibly LTBI ("OD"), TB from Human Immunodeficiency Virus ("HIV"), TB from pneumonia ("PNA"), or identify risk of progression to TB ("RISK"), risk of TB treatment failure ("FAIL"), or classify treatment responses (i.e., failures from cures, "RES").

The tissue type variable denotes whether the signature was developed using samples of either whole blood/paxgene or peripheral blood mononuclear cells (PBMCs). Due to the manipulation of cells inherently required to obtain PBMCs, many scientists prefer to use only whole blood samples for analysis.
  
```{r descriptive objs}
# Create annotation object ----------------------------------------------------

sigAnnotData <- read.csv("C:/Users/aubre/Documents/BU Research/Tuberculosis/TBSP Development/sig_annotations.csv")[,1:3]
colnames(sigAnnotData) <- c("names", "disease", "tissue_type")
sigAnnotData$disease <- revalue(sigAnnotData$disease, 
                          replace = c("Disease" = "Disease",
                                      "Disease/HIV" = "HIV",
                                      "Disease/Other Diseases"= "OD",
                                      "Disease/Pneumonia" = "PNA",
                                      "failure" = "FAIL",
                                      "response" = "RES", "risk" = "RISK"))
sigAnnotData$tissue_type <- revalue(sigAnnotData$tissue_type,
                               replace = c("whole blood" = "whole"))
sigAnnotData$names <- names(TBsignatures)
save(sigAnnotData, file = "data/sigAnnotData.rda")

common_sigAnnotData <- sigAnnotData
common_sigAnnotData$names <- names(TBcommon)
save(common_sigAnnotData, file = "data/common_SigAnnotData.rda")

```

# Add session info